plugins/zoom-to-fit/src/index.js

```diff
@@ -0,0 +1,230 @@
+  }
+
+  /**
+   * Creates DOM for ui element.
+   * @private
+   */
+  createDom_() {
+    // const parentSvg = this.workspace_.getParentSvg();
+    // this.svgGroup_ = Blockly.utils.dom.createSvgElement(
+    //     Blockly.utils.Svg.G, {'class': 'zoomToFit'}, parentSvg);
+
+    this.svgGroup_ = Blockly.utils.dom.createSvgElement(
+        Blockly.utils.Svg.RECT,
+        {
+          'class': 'zoomToFit',
+          'width': this.WIDTH_,
+          'height': this.HEIGHT_,
+        },
+        null);
+    Blockly.utils.dom.insertAfter(
+        this.svgGroup_, this.workspace_.getBubbleCanvas());
+
+    // Attach listener.
+    this.onZoomOutWrapper_ = Blockly.browserEvents.conditionalBind(
+        this.svgGroup_, 'mousedown', null, this.onClick_.bind(this));
+  }
+
+  /**
+   * Handle click event.
+   * @private
+   */
+  onClick_() {
+    this.workspace_.zoomToFit();
+    const uiEvent = new (Blockly.Events.get(Blockly.Events.CLICK))(
+        null, this.workspace_.id, 'zoom_reset_control');
+    Blockly.Events.fire(uiEvent);
+  }
+
+
+  /**
+   * Returns the bounding rectangle of the UI element in pixel units relative to
+   * the Blockly injection div.
+   * @return {!Blockly.utils.Rect} The plugin’s bounding box.
+   */
+  getBoundingRectangle() {
+    return new Blockly.utils.Rect(
+        this.top_, this.top_ + this.HEIGHT_,
+        this.left_, this.left_ + this.WIDTH_);
+  }
+
+  /**
+   * Positions the zoom-to-fit control.
+   * It is positioned in the opposite corner to the corner the
+   * categories/toolbox starts at.
+   * @param {!Blockly.MetricsManager.UiMetrics} metrics The workspace metrics.
+   * @param {!Array<Blockly.utils.Rect>} savedPositions List of rectangles that
+   *     are already on the workspace.
+   */
+  position(metrics, savedPositions) {
+    // Not yet initialized.
+    if (!this.verticalSpacing_) {
+      return;
+    }
+    if (metrics.toolboxMetrics.position === Blockly.TOOLBOX_AT_LEFT ||
+        (this.workspace_.horizontalLayout && !this.workspace_.RTL)) {
+      // Right corner placement.
+      this.left_ = metrics.viewMetrics.width + metrics.absoluteMetrics.left -
+          this.WIDTH_ - this.MARGIN_SIDE_ -
+          Blockly.Scrollbar.scrollbarThickness;
+    } else {
+      // Left corner placement.
+      this.left_ = this.MARGIN_SIDE_ + Blockly.Scrollbar.scrollbarThickness;
+    }
+
+    // Upper corner placement
+    const minTop = this.top_ =
+        metrics.absoluteMetrics.top + this.verticalSpacing_;
+    // Bottom corner placement
+    const maxTop = metrics.absoluteMetrics.top + metrics.viewMetrics.height -
+        this.HEIGHT_ - this.verticalSpacing_;
+    const placeBottom =
+        metrics.toolboxMetrics.position !== Blockly.TOOLBOX_AT_BOTTOM;
+    this.top_ = placeBottom ? maxTop : minTop;
+
+    // Check for collision and bump if needed.
+    let boundingRect = this.getBoundingRectangle();
+    for (let i = 0, otherEl; (otherEl = savedPositions[i]); i++) {
```

Comment:
Can this ever loop infinitely, in a small enough workspace?

Reply:
No. It is either always bumped up or always bumped down.

Reply:
Right, but after a bump you reset the index to -1 to restart the loop. So if on the first run it has to bump up, and then on the second try it overlaps the other direction and bumps down, it will keep looping.

Reply:
What I meant by always bumped up or down is that the `bumpUp` boolean is never changed between loops.
It's set before the loop only once.